### PR TITLE
feat(render-engine): move cache to viewport

### PR DESCRIPTION
**Pull request by lumixraku**
A description of the proposed changes:
Moving the cache canvas from the spreadsheet to the viewport is primarily for the scenario of freezing rows and columns.

How to test them:
- Performance:  The FPS performance during scrolling should theoretically be consistent with the performance when there are no frozen rows or columns.

- Font rendering: specially in scenarios involving overflow and merged cells spanning frozen rows and columns.

- Correct texture mapping:  such as avoiding blank textures and ensuring complete rendering of text and background colors.

BREAKING CHANGE:
Before:
cache canvas was controlled by spreadsheet.
After: 
cache canvas was created by viewport

### CODE REVIEW

**Comment by DR-Univer** [comment]
slide dev可以展示，这个分支白屏了，得看看。

**Comment by DR-Univer** [comment]
image
文本滚动后会被加粗，点击canvas后消失

File: packages/engine-render/src/basics/vector2.ts [hunk 1]
```diff
@@ -844,17 +846,64 @@ export interface IBoundRect {
 }
 
 export interface IBoundRectNoAngle {
+    /**
+     * 冻结区域相对 MainCanvas 的物理位置,
```
Comment (Jocs) [inline packages/engine-render/src/basics/vector2.ts:850]: English prefer

**Comment by lumixraku** [comment]
> slide dev可以展示，这个分支白屏了，得看看。

rrelativeXY, fixed

> image 文本滚动后会被加粗，点击canvas后消失

DPI 和分辨率多少?  建议设置并没有出现,  我都调整为不建议也没出现

File: packages/engine-render/src/viewport.ts [hunk 2]
```diff
@@ -156,10 +168,42 @@ export class Viewport {
...
+     */
+    private _isDirty = true;
+    private _cacheCanvas: UniverCanvas | null = null;
+
+    /**
+     * Allow Caching: Default is true.
+     * The configuration comes from the props.allowCache passed in during viewport initialization.
+     * When _allowCache is true, a cacheCanvas will be created.
+     */
+    private _allowCache: boolean = true;
```
Comment (DR-Univer) [inline packages/engine-render/src/viewport.ts:197]: 默认开启，是否doc或slide要主动去关闭。

File: packages/sheets-ui/src/controllers/sheet-render.controller.ts [hunk 3]
```diff
@@ -190,9 +196,99 @@ export class SheetRenderController extends RxDisposable {
...
                     if (renderer.mainComponent instanceof Spreadsheet) {
-                        renderer.mainComponent.makeForceDirty(true);
+                        (renderer.mainComponent as Spreadsheet).makeForceDirty(true);
                     }
                 });
             });
     }
+
+    private _spreadsheetViewports(scene: Scene) {
+        return scene.getViewports().filter((v) => ['viewMain', 'viewMainLeftTop', 'viewMainTop', 'viewMainLeft'].includes(v.viewportKey));
```
Comment (DR-Univer) [inline packages/sheets-ui/src/controllers/sheet-render.controller.ts:206]: 有专门针对viewport的常量，例如VIEW_MAIN，文本尽量都用常量

Comment (lumixraku) [reply packages/sheets-ui/src/controllers/sheet-render.controller.ts:206]: 不过常量都定义在 sheet-ui 中,  engine-render 也可以引用吗?

File: packages/engine-render/src/layer.ts [hunk 4]
```diff
@@ -241,15 +243,23 @@ export class Layer extends Disposable {
...
     private _draw(mainCtx: UniverRenderingContext, isMaxLayer: boolean) {
-        this._scene.getViewports()?.forEach((vp) => vp.render(mainCtx, this.getObjectsByOrder(), isMaxLayer));
+        const viewports = this._scene.getViewports().filter((vp) => vp.shouldIntoRender());
+        const objects = this.getObjectsByOrder();
+        for (const [_index, vp] of viewports.entries()) {
+            vp.render(mainCtx, objects, isMaxLayer);
+        }
+        objects.forEach((o) => {
+            o.makeDirty(false);
+            (o as unknown as { makeForceDirty?: (state: boolean) => void }).makeForceDirty?.(false);
```
[the remark this answers is not in the document]
Comment (lumixraku) [reply packages/engine-render/src/layer.ts:261]: 是, base 没有 forceDirty.  OK

File: packages/engine-render/src/viewport.ts (hunk 2 above)
Comment (lumixraku) [reply packages/engine-render/src/viewport.ts:197]: change back to default: false

File: packages/core/src/shared/tools.ts [hunk 5]
```diff
@@ -671,4 +671,35 @@ export class Tools {
 
         return !containsInvalidChars && isValidLength;
     }
+
+    static hasIntersectionBetweenTwoBounds(
```
[the remark this answers is not in the document]
Comment (lumixraku) [reply packages/core/src/shared/tools.ts:675]: OK, 不过之前只有 range 相交判断, 没有 bound 相交判断

File: packages/engine-render/src/components/sheets/spreadsheet.ts [hunk 6]
```diff
@@ -220,138 +224,209 @@ export class Spreadsheet extends SheetComponent {
...
             } else {
-                this._draw(mainCtx, bounds);
+                this._draw(mainCtx, viewportInfo);
             }
+        } else if (this.sheetHeaderViewport().includes(viewPortKey)) {
+            // doing nothing, other components(SpreadsheetRowHeader...) will render
         } else {
-            this._draw(mainCtx, bounds);
+            // embed in doc & slide
+            this._draw(mainCtx, viewportInfo);
```
[the remark this answers is not in the document]
Comment (lumixraku) [reply packages/engine-render/src/components/sheets/spreadsheet.ts:407]: 这里和目前 dev 分支的逻辑保持一致,  dev 中 slides 也是走的 draw, 没有贴图.

File: packages/sheets-ui/src/controllers/sheet-render.controller.ts (hunk 3 above)
Comment (lumixraku) [reply packages/sheets-ui/src/controllers/sheet-render.controller.ts:206]: 我计划都移动到 engine 里

Comment (DR-Univer) [reply packages/sheets-ui/src/controllers/sheet-render.controller.ts:206]: 可以的 ，让engine-render 统一管理这些名称，后续sheet doc slide 都采取这个策略

File: packages/core/src/shared/tools.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DR-Univer) [reply packages/core/src/shared/tools.ts]: 可以新增一个，说明一下用途就行，方便后人复用

**Review by DR-Univer — APPROVED** [review]

**Pull request merged**